packages/typescript-estree/src/create-program/getParsedConfigFile.ts

```diff
@@ -0,0 +1,64 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import type * as ts from 'typescript/lib/tsserverlibrary';
+
+import { CORE_COMPILER_OPTIONS } from './shared';
+
+/**
+ * Utility offered by parser to help consumers parse a config file.
+ *
+ * @param configFile the path to the tsconfig.json file, relative to `projectDirectory`
+ * @param projectDirectory the project directory to use as the CWD, defaults to `process.cwd()`
+ */
+function getParsedConfigFile(
+  configFile: string,
+  projectDirectory?: string,
+): ts.ParsedCommandLine {
```

Comment:
[Refactor] I'm not a fan of functions throwing errors as a form of control flow. It can't be described in the function's signature and so can easily become hard to keep track of as functions get more complex.

We're seeing an example of here. The thrown error deeper in `getParsedConfigFile` is caught and used to throw a new message in `createProjectService`. I find that hard to keep track of.

What do you think about having this function instead return a `string` message to indicate there was a failure parsing?

I.e.:

```suggestion
): ts.ParsedCommandLine | string {
```

...and then later, instead of `throw new Error(...)`s, it'd return the string message. `onUnRecoverableConfigFileDiagnostic` could set `parsed` to be a `string`:

```ts
let parsingError: string | undefined;

const parsed = tsserver.getParsedCommandLineOfConfigFile(
  // ...
      onUnRecoverableConfigFileDiagnostic: diag => {
        parsingError = formatDiagnostics([diag]);
      },
  // ...
);
```

(or another strategy like that, e.g. having `let parsed: string | ts.ParsedCommandLine`)

Reply:
I'll get this change in.  I typically use an Either monad to keep errors in the type system, so the objective works for me.  Good call.